File: app/models/workflows.py
# app/models/workflows.py
from __future__ import annotations
from typing import Optional, List, Dict, Any
from pydantic import BaseModel, Field


class WorkflowStep(BaseModel):
    action_code: str
    label: Optional[str] = None
    input_required: bool = False
    output_key: Optional[str] = None
    ui_type: str = "list"
    params: Dict[str, Any] = Field(default_factory=dict)

    skipped: bool = False
    auto_assign: bool = False


class WorkflowDefinition(BaseModel):
    tenant: str
    workflow_id: str
    name: str
    steps: List[WorkflowStep]
    active: bool = True
    requires_caps: List[str] = Field(default_factory=list)
    created_at: Optional[Any] = None
    updated_at: Optional[Any] = None


class WorkflowActionMeta(BaseModel):
    action_code: str
    label: str
    input_required: bool
    output_key: Optional[str] = None
    ui_type: str
    description: Optional[str] = None
    module: Optional[str] = None
    group: Optional[str] = None
    requires_caps: List[str] = Field(default_factory=list)

File: admin_ui/src/api/workflows.ts
import { api } from './axios'

export type WorkflowStep = {
  action_code: string
  label?: string
  input_required: boolean
  output_key?: string
  ui_type: string
  params?: any
}

export type WorkflowDefinition = {
  tenant: string
  workflow_id: string
  name: string
  steps: WorkflowStep[]
  active: boolean
}

export type WorkflowActionMeta = {
  action_code: string
  label: string
  input_required: boolean
  output_key?: string
  ui_type: string
  description?: string
  module?: string
  group?: string
}

export async function listWorkflows(tenant: string): Promise<{ items: WorkflowDefinition[] }> {
  const res = await api.get(`/tenants/${tenant}/workflows`)
  return res.data
}

export async function getWorkflow(tenant: string, id: string): Promise<WorkflowDefinition> {
  const res = await api.get(`/tenants/${tenant}/workflows/${id}`)
  return res.data
}

export async function upsertWorkflow(tenant: string, wf: WorkflowDefinition) {
  const res = await api.post(`/tenants/${tenant}/workflows`, wf)
  return res.data
}

export async function deleteWorkflow(tenant: string, workflowId: string): Promise<{ ok: boolean }> {
  const res = await api.delete(`/tenants/${encodeURIComponent(tenant)}/workflows/${encodeURIComponent(workflowId)}`)
  return res.data
}

export async function listAvailableWorkflowActions(tenant?: string): Promise<{ items: WorkflowActionMeta[] }> {
  const res = await api.get('/workflows/available-actions', { params: tenant ? { tenant } : undefined })
  return res.data
}

export type WorkflowAuditItem = {
  tenant: string
  workflow_id: string
  name: string
  valid: boolean
  errors: string[]
  repair_available: boolean
  repair_notes: string[]
  errors_after_repair: string[]
}

export async function auditWorkflows(tenant: string): Promise<{ items: WorkflowAuditItem[] }> {
  const res = await api.get(`/tenants/${tenant}/workflows/audit`)
  return res.data
}

export async function repairWorkflows(tenant: string) {
  const res = await api.post(`/tenants/${tenant}/workflows/repair`)
  return res.data
}
